### PR TITLE
[unitaryhack] a quantum simulator in PyTorch #1225

### PR DESCRIPTION
**Pull request by arshpreetsingh**
**Context:**

A quantum simulator in PyTorch #1225

**Description of the Change:**
1. Added new utility module for Pytorch support. (pennylane/devices/pt_ops.py)
2. Added Interface Class DefaultQubitTF(DefaultQubit). (pennylane/devices/default_qubit_tf.py)
3. Registered for DefaultQbit. (pennylane/devices/default_qubit.py)
4. Unit-Test: (tests/devices/test_default_qubit_pt.py)
5. Unit-Test (tests/devices/test_default_qubit_pt.py)

**Benefits:**
1. The entire computation is end-to-end differentiable using native backpropagation,
2. The computation can take advantage of specific runtime features of the framework, such as GPU/TPU support, XLA acceleration, etc.

**Possible Drawbacks:**

**Related GitHub Issues:**

https://github.com/PennyLaneAI/pennylane/issues/1225

**Pending Items:**

1. Require more information for tf.linalg.diag() alternate for def CRot(a, b, c) in  (pennylane/devices/pt_ops.py)
2. Require support for Function-Doc pennylane/devices/default_qubit_pt.py
3. Require support for  tf.GradientTape() , What alternative can be used for Pytorch. --> (tests/devices/test_default_qubit_pt.py)

### CODE REVIEW

File: pennylane/devices/default_qubit_pt.py [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+
+    C_DTYPE = torch.complex128
+    R_DTYPE = torch.float64
+    _asarray = staticmethod(torch.as_tensor)
+    _dot = staticmethod(lambda x, y: torch.tensordot(x, y, axis=1))
+    _abs = staticmethod(torch.abs)
+    _reduce_sum = staticmethod(torch.sum)
+    _reshape = staticmethod(torch.reshape)
+    ## TODO:  Need to verify reshape
+    _flatten = staticmethod(lambda tensor: torch.reshape(tensor, [-1]))
```
Comment (josh146) [inline pennylane/devices/default_qubit_pt.py:139]: I believe the torch version is `torch.view`?

File: pennylane/devices/default_qubit_pt.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+            passthru_interface="pt",
+            supports_reversible_diff=False,
+        )
+        return capabilities
+
+    ## TODO:  this needs to be updated
+    # @staticmethod
+    # def _scatter(indices, array, new_dimensions):
+    #     indices = np.expand_dims(indices, 1)
+    #     return tf.scatter_nd(indices, array, new_dimensions)
```
Comment (josh146) [inline pennylane/devices/default_qubit_pt.py:183]: Something like this _might_ work, but I have not tested this or verified it :)

```python
def _scatter(indices, array, new_dimensions):
    new_tensor = torch.zeros(new_dimensions, dtype=array.dtype)
    new_tensor[indices] = array
    return tensor
```

**Comment by josh146** [comment]
Thanks for this nice contribution @arshpreetsingh! Let us know how you are going, and feel free to ask us any questions in the PR itself :)

File: pennylane/devices/default_qubit_pt.py (hunk 2 above)
Comment (arshpreetsingh) [reply pennylane/devices/default_qubit_pt.py]: kind of stuck on this one, tried following:

```
import torch
indices = torch.tensor([[4], [3], [1], [7]])
updates = torch.tensor([9,10,11,12])
shape2 = (8,0)
new_tensor = torch.zeros((8,0), dtype=updates.dtype)
new_tensor[indices] = updates
print(new_tensor)
```
RuntimeError: shape mismatch: value tensor of shape [4] cannot be broadcast to indexing result of shape [4, 1, 0]

**Comment by arshpreetsingh** [comment]
> Thanks for this nice contribution @arshpreetsingh! Let us know how you are going, and feel free to ask us any questions in the PR itself :)



> Thanks for this nice contribution @arshpreetsingh! Let us know how you are going, and feel free to ask us any questions in the PR itself :)

just wanted to know about tf.GradientTape()'s Alternate, There are some discussions in the Pytorch side here, Not sure either it is available or not.
https://github.com/pytorch/pytorch/issues/23475

**Comment by josh146** [comment]
> just wanted to know about tf.GradientTape()'s Alternate, There are some discussions in the Pytorch side here, Not sure either it is available or not.

Is this for writing tests? If so, there is no direct `tf.GradientTape()` analogue in PyTorch. Instead, gradients can be computed like so:

```python
weights = torch.tensor([0.1, 0.2, 0.3])
loss = torch.sum(torch.cos(weights))
loss.backward()  # compute the backward pass
weights.grad  # the gradient of the weights vector
```

You can see an example of this in the `test_qnode_torch.py` test file:

https://github.com/PennyLaneAI/pennylane/blob/a9a914d9b6e475374dba1e91a22f2635b97df913/tests/interfaces/test_qnode_torch.py#L199

**Comment by josh146** [comment]
Hey @arshpreetsingh, just checking in to see how everything is going, and if there is anything in particular you have questions about or want a hand with :)

**Comment by arshpreetsingh** [comment]
> Hey @arshpreetsingh, just checking in to see how everything is going, and if there is anything in particular you have questions about or want a hand with :)

Hi @josh146 , I got little side-tracked yesterday while resolving test-Cases, As I understand now each default-qbit-device(Tensorflow, Pytorch etc ) should have independent private functions like _asarry, _cast to perform operation related to Quantum-Gates or Neural-Network Operations, I also found Slimane33 has better ideas introduced in this PR: https://github.com/PennyLaneAI/pennylane/pull/1360 like GPU support as well. I was Planning to abandon my PR and  contribute in  Slimane33's PR(like performing testCases and all) Looking forward for your idea on this.

**Comment by josh146** [comment]
> I was Planning to abandon my PR and contribute in Slimane33's PR(like performing testCases and all) Looking forward for your idea on this.

From my understanding teams are allowed (and encouraged!) for unitaryHACK, so definitely feel free to discuss with @Slimane33 :)

**Comment by Slimane33** [comment]
Hi @arshpreetsingh you're welcome to contribute! Feel free to send me a PM on discord on the UnitaryHACK server :)

**Comment by arshpreetsingh** [comment]
> Hi @arshpreetsingh you're welcome to contribute! Feel free to send me a PM on discord on the UnitaryHACK server :)

Great @Slimane33  , Sent you friend request there at discord, Waiting for connection.

**Pull request closed without merging**